### PR TITLE
COS-8: Dynamically render form config (no formatting/styling/input capture)

### PR DESCRIPTION
**Pull request by RussianPrussian**
This PR introduces a component that includes the ability to transform a category form config (json) to a typescript object and to subsequently render the encapsulated info as HTML. 

There will be follow up stories to do the following:
1. Connect HTTP service logic to this function
2. Introduce ability to actually capture the input fields and POST that info to the backend
3. Format and style the rendered HTML using Bootstrap.

@tariqbroadnax is currently working on adding some tests.

### CODE REVIEW

File: src/app/dashboard/expense-category-input-renderer/expense-category-json-transformer.service.spec.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    expect(childFieldGroups[0].expenseTagList[0] instanceof InputTag).toBeTruthy();
+    expect(childFieldGroups[0].expenseTagList[0] instanceof TextAreaTag).toBeFalsy();
+  });
+
+  it('should throw an error if the converted JSON object do not have the required label and/or tag', () => {
+    expect(() => service.transformJson(sample1)).toThrowError();
+  });
+
+  it('should successfully transform a JSON object that  does not have any fields', () => {
+    expect(service.transformJson(sample2)).toBeTruthy();
```
Comment (RussianPrussian) [inline src/app/dashboard/expense-category-input-renderer/expense-category-json-transformer.service.spec.ts]: this needs to actually check that it doesn't have fields.

File: src/app/dashboard/expense-category-input-renderer/expense-category-json-transformer.service.spec.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  it('should throw an error if the converted JSON object do not have the required label and/or tag', () => {
+    expect(() => service.transformJson(sample1)).toThrowError();
+  });
+
+  it('should successfully transform a JSON object that  does not have any fields', () => {
+    expect(service.transformJson(sample2)).toBeTruthy();   
+  });
+
+  it('should successfully transform a JSON object with multiple levels of child field groups', () => {
+    expect(service.transformJson(sample3)).toBeTruthy();
```
Comment (RussianPrussian) [inline src/app/dashboard/expense-category-input-renderer/expense-category-json-transformer.service.spec.ts]: should actually perform checks on the fields.

File: src/app/dashboard/expense-category-input-renderer/expense-category-json-transformer.service.spec.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    expect(config.expenseTagList[0].attrs["min"]).toBe("2009-10-13");
+    expect(childFieldGroups.length).toBeGreaterThan(0);
+    expect(childFieldGroups[0].tagGroupLabel).toBe("Hotel");
+    expect(childFieldGroups[0].expenseTagList[0].label).toBe("Hotel Name");
+    expect(childFieldGroups[0].expenseTagList[0] instanceof InputTag).toBeTruthy();
+    expect(childFieldGroups[0].expenseTagList[0] instanceof TextAreaTag).toBeFalsy();
+  });
+
+  it('should throw an error if the converted JSON object do not have the required label and/or tag', () => {
+    expect(() => service.transformJson(sample1)).toThrowError();
```
Comment (RussianPrussian) [inline src/app/dashboard/expense-category-input-renderer/expense-category-json-transformer.service.spec.ts]: name samples as something more descriptive ... not just sample1 sample2 etc.

File: src/app/dashboard/expense-category-input-renderer/expense-category-json-transformer.service.spec.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    expect(childFieldGroups[0].expenseTagList[0] instanceof TextAreaTag).toBeFalsy();
+  });
+
+  it('should throw an error if the converted JSON object do not have the required label and/or tag', () => {
+    expect(() => service.transformJson(configNoLabelTag)).toThrowError();
+  });
+
+  it('should successfully transform a JSON object that does not have any fields', () => {
+    let config: ExpenseCategoryConfig = service.transformJson(configNoFields);
+    expect(config.expenseTagList.length).toBe(0);
```
Comment (RussianPrussian) [inline src/app/dashboard/expense-category-input-renderer/expense-category-json-transformer.service.spec.ts]: can you actually check some of the elements?

File: src/app/dashboard/expense-category-input-renderer/expense-category-json-transformer.service.spec.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  });
+
+  it('should successfully transform a JSON object with multiple levels of child field groups', () => {
+    let config: ExpenseCategoryConfig = service.transformJson(configNested);
+    expect(config.expenseTagGroups[0].expenseTagGroups).toBeTruthy();
```
Comment (RussianPrussian) [inline src/app/dashboard/expense-category-input-renderer/expense-category-json-transformer.service.spec.ts]: check soem elements

File: src/app/dashboard/expense-category-input-renderer/expense-category-json-transformer.service.spec.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  });
+
+  it('should throw an error if a JSON object has multiple fields have the same ID', () => {
+    expect(() => service.transformJson(configSameId)).toThrowError();
```
Comment (RussianPrussian) [inline src/app/dashboard/expense-category-input-renderer/expense-category-json-transformer.service.spec.ts]: we should have some tests for other tag types.

File: src/app/dashboard/expense-category-input-renderer/expense-category-json-transformer.service.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    let expenseCatConfig = new ExpenseCategoryConfig();
+    expenseCatConfig.tagGroupLabel = rawJsonObject["fieldGroupLabel"];
+    expenseCatConfig.expenseTagList = new Array<GeneratedTag>();
+
+    //Construct field object based on tag type (discriminated by the "tag" field)
+    let ids: Array<number> = new Array<number>();
+    for(let i in rawJsonObject["fieldArray"]){
+      let rawFieldElement = rawJsonObject["fieldArray"][i];
+
+      if(ids.includes(rawFieldElement["id"])) {
```
Comment (RussianPrussian) [inline src/app/dashboard/expense-category-input-renderer/expense-category-json-transformer.service.ts]: extract into its own method called "verifyUniqueId"

File: src/app/dashboard/expense-category-input-renderer/expense-category-json-transformer.service.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+import { TextAreaTag } from './text-area-tag';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ExpenseCategoryJSONTransformerService {
+
+  private static tagTypeMap = new Map<string, (fieldProps: GeneratedTag) => {}>();
+
+  constructor(private http:HttpClient) {
```
Comment (NateHand93) [inline src/app/dashboard/expense-category-input-renderer/expense-category-json-transformer.service.ts]: HttpClient injected but never used?

Comment (RussianPrussian) [reply src/app/dashboard/expense-category-input-renderer/expense-category-json-transformer.service.ts]: addressed.

Comment (RussianPrussian) [reply src/app/dashboard/expense-category-input-renderer/expense-category-json-transformer.service.ts]: wtf! still here.

Comment (RussianPrussian) [reply src/app/dashboard/expense-category-input-renderer/expense-category-json-transformer.service.ts]: Okay. actually gone now.

File: src/app/dashboard/expense-category-input-renderer/expense-category-json-transformer.service.ts (hunk 7 above)
Comment (RussianPrussian) [reply src/app/dashboard/expense-category-input-renderer/expense-category-json-transformer.service.ts]: @tariqbroadnax I feel like you did this, but somehow it got reverted.

**Pull request merged**